add a new at::lift operator, fix torch.tensor for functionalization

bdhirsh opened the pull request:
Re-land of https://github.com/pytorch/pytorch/pull/76319 - I needed to tell `onnx` what the new `at::lift` op is, since it technically gets traced by `torch.jit.trace()` into torchscript. I think it should be a no-op though.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77285

This reverts commit 85bd65a880ddd7a1f4b1ea4288423d75d45a56b3.

bdhirsh committed:
add a new at::lift operator, fix torch.tensor for functionalization

This reverts commit 85bd65a880ddd7a1f4b1ea4288423d75d45a56b3.

[ghstack-poisoned]

bdhirsh commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -3927,6 +3927,9 @@ def linspace(g, start, end, steps, dtype, layout, device, pin_memory):
 
+def lift(g, self):
+    # at::lift() is a no-op from the perspective of tracing for onnx
+    return self
```
After reading through some onnx docs / looking at previous PR's, I added this and was able to get the failing onnx tests to pass locally. (I think a no-op is fine since lifting should be traced away before you create a graph)

bdhirsh replied:
(this is the only change on this PR compared to the original)

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3328,5 +3328,12 @@ at::Tensor diagonal_scatter(const at::Tensor& self, const at::Tensor& src, int64
+// The default implementation of lift is a no-op.
+// If TLS is set appropriately (for wrapper-tensor keys like Functionalize or functorch transforms),
+// then we'll dispatch to one of their implementations, which will properly lift the tensor into a wrapper.
+at::Tensor lift(const at::Tensor& self) {
```
In which case can we actually reach this implementation?
Should this be an `TORCH_INTERNAL_ASSERT(false, ...)` ?

bdhirsh replied:
This'll actually get called in all other cases from `tensor_new.cpp`. when we `return tensor.lift()`, it'll dispatch to `lift_functional()` if functionalization is enabled (or eventually, if the device is Lazy/XLA), and otherwise dispatch and and no-op.

albanD reviewed (commented):
Sounds good for a reland. Shouls we add `ciflow/trunk` to get full signal on this PR?

bdhirsh committed:
Update on "add a new at::lift operator, fix torch.tensor for functionalization"

Re-land of https://github.com/pytorch/pytorch/pull/76319 - I needed to tell `onnx` what the new `at::lift` op is, since it technically gets traced by `torch.jit.trace()` into torchscript. I think it should be a no-op though.


This reverts commit 85bd65a880ddd7a1f4b1ea4288423d75d45a56b3.

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge this please
